rs-backend/src/load.rs

```diff
@@ -103,46 +99,54 @@ impl InputData {
                 continue;
             }
 
-            let header = InputHeader {
-                commit: contents.lookup("header.commit").unwrap().as_str().unwrap().to_string(),
-                date: InputData::parse_from_header(contents.lookup("header.date")
-                        .unwrap()
-                        .as_str()
-                        .unwrap()).or_else(|_| InputData::parse_from_filename(&filename))?,
-            };
-            let date = header.date;
+            let commit = contents.lookup("header.commit").unwrap().as_str().unwrap().to_string();
+            let date = InputData::parse_from_header(contents.lookup("header.date")
+                    .unwrap()
+                    .as_str()
+                    .unwrap()).or_else(|_| InputData::parse_from_filename(&filename))?;
 
             let test_name = filename[0..filename.find("--").unwrap()].to_string();
 
             let times = contents.find("times").unwrap().as_array().unwrap();
-            if &test_name == "rustc" {
-                data_rustc.push(TestRun::new(date, header, times, test_name));
+
+            let push_to = if &test_name == "rustc" {
+                &mut data_rustc
             } else {
-                let index = data_benchmarks.iter()
-                    .position(|benchmark: &TestRun| benchmark.date == date);
-                if let Some(index) = index {
-                    c_benchmarks_add += 1;
-                    let crate_name = times[0].find("crate").unwrap().as_str().unwrap();
-                    data_benchmarks[index].by_crate.insert(test_name.to_string(),
-                                                           make_times(times, false)
-                                                               .remove(crate_name)
-                                                               .unwrap());
-                } else {
-                    data_benchmarks.push(TestRun::new(date, header, times, test_name));
+                &mut data_benchmarks
+            };
+
+            // A run on the same day occurred. Merge the two data sets,
+            // overwriting previous data if the same crate was benchmarked.
+            if let Some(index) = push_to.iter().position(|run: &TestRun| run.date == date) {
+                let run = &mut push_to[index];
```

Comment:
If there's a way to use `iter_mut().find(...)` here we should go for it, but I think non-lexical borrow are needed for that; I may be wrong though.